### PR TITLE
Fixes part of #16 and #44: Implementation of ProfileChooserFragment

### PR DESCRIPTION
**Pull request by jamesxu0**
## Explanation

Fixes part of #16 and #44: Implementation of ProfileChooserFragment. It also introduces ProfileTestHelper to help other test cases initialize admin and user profiles. Note the base branch is not develop but refers to PR #269.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -74,6 +74,7 @@ dependencies {
       'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.2.1',
       'org.jetbrains.kotlinx:kotlinx-coroutines-android:1.2.1',
       'org.mockito:mockito-core:2.7.22',
+      'de.hdodenhof:circleimageview:3.0.1'
```
Comment (rt4914) [inline app/build.gradle]: Nit: Add libraries in alphabetical order.

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -15,11 +15,13 @@
     <activity android:name=".testing.ContinuePlayingFragmentTestActivity" />
     <activity android:name=".home.HomeActivity" />
     <activity android:name=".player.audio.testing.AudioFragmentTestActivity" />
+    <activity
```
Comment (rt4914) [inline app/src/main/AndroidManifest.xml]: Nit: Keep activities in alphabetical order here.

File: app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt [hunk 3]
```diff
@@ -48,3 +48,11 @@ fun setImageDrawable(imageView: ImageView, thumbnailGraphic: LessonThumbnailGrap
     }
   )
 }
+
+@BindingAdapter("profile:src")
+fun setProfileImage(imageView: ImageView, imageUrl: String) {
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt:76]: check function at top of this class `setImageDrawable`, I think that can be used here.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package org.oppia.app.databinding
+
+import android.text.format.DateUtils
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.databinding.BindingAdapter
+import java.util.*
+
+/** Converts time in ms to readable relative time string. */
+@BindingAdapter("profile:date")
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Suggestion: keep this `android:date`

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+fun setTextWithDate(textView: TextView, timeMs: Long) {
+  val dateText = "Last used " + DateUtils.getRelativeTimeSpanString(timeMs, Date().time, DateUtils.MINUTE_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE)
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: `option + cmd + L` to reorganise this line.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.oppia.app.databinding
+
+import android.text.format.DateUtils
+import android.widget.ImageView
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Remove this unused import.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 7]
```diff
@@ -1,29 +1,92 @@
...
+import org.oppia.app.home.HomeActivity
+import org.oppia.app.model.ProfileChooserModel
+import org.oppia.app.recyclerview.BindableAdapter
 import org.oppia.app.viewmodel.ViewModelProvider
+import org.oppia.domain.profile.ProfileManagementController
 import javax.inject.Inject
 
 /** The presenter for [ProfileChooserFragment]. */
 @FragmentScope
 class ProfileChooserFragmentPresenter @Inject constructor(
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:23]: `option + cmd + L` to reorganise this file.

Comment (rt4914) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:23]: Ditto for this and other files too.

File: app/src/main/java/org/oppia/app/splash/SplashActivity.kt [hunk 8]
```diff
@@ -5,17 +5,17 @@ import android.os.Bundle
...
 
-/** An activity that shows a temporary loading page until the app is fully loaded then navigates to [HomeActivity]. */
+/** An activity that shows a temporary loading page until the app is fully loaded then navigates to [ProfileActivity]. */
 class SplashActivity : AppCompatActivity() {
 
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
     setContentView(R.layout.splash_activity)
 
     getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
```
Comment (rt4914) [inline app/src/main/java/org/oppia/app/splash/SplashActivity.kt]: Replace `getWindow()` to `window`

File: app/src/main/res/values/colors.xml [hunk 9]
```diff
@@ -23,6 +23,9 @@
   <color name="black">#000000</color>
   <!-- AUDIO COMPONENT -->
   <color name="audioComponentBackground">@color/oppiaDarkBlue</color>
+  <!-- PROFILE COMPONENT -->
+  <color name="avatarBorder">#4E4E50</color>
+  <color name="chooserBackground">#212121</color>
```
Comment (rt4914) [inline app/src/main/res/values/colors.xml]: suggest renaming this to `profileChooserBackground`

File: app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+      onView(atPositionOnView(R.id.profile_recycler_view, 0, R.id.profile_is_admin_text)).check(matches(withText(context.getString(R.string.profile_chooser_admin))))
+      onView(atPositionOnView(R.id.profile_recycler_view, 1, R.id.profile_name_text)).check(matches(withText("Ben")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 1, R.id.profile_is_admin_text)).check(matches(not(isDisplayed())))
+      onView(atPositionOnView(R.id.profile_recycler_view, 2, R.id.add_profile_text)).check(matches(withText(context.getString(R.string.profile_chooser_add))))
+    }
+  }
+
+  @Test
+  @ExperimentalCoroutinesApi
+  fun testProfileChooserFragment_addManyProfiles_checkProfilesSortedAndNoAddProfile() {
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:72]: Test case is failing with following error:
Screenshot 2019-11-13 at 12 27 06 PM

Comment (rt4914) [reply app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:72]: This might be because of corouties too.

**Review by rt4914 — CHANGES_REQUESTED** [review]
LGTM, there is one test case which is failing which might be because of corouties too. Reassign it to me once you have addressed all the comments.

File: app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt (hunk 3 above)
Comment (BenHenning) [reply app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt:76]: Don't we need this for URL loading, though?

File: app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt [hunk 11]
```diff
@@ -48,3 +48,11 @@ fun setImageDrawable(imageView: ImageView, thumbnailGraphic: LessonThumbnailGrap
...
+  Glide.with(imageView.context)
+    .load(imageUrl)
+    .placeholder(R.drawable.default_avatar)
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt]: Reminder to pull in a vector version of this instead of PNG.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import android.widget.TextView
+import androidx.databinding.BindingAdapter
+import java.util.*
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Please don't widlcard import since it pollutes the symbol namespace.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt (hunk 4 above)
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Are we using this anywhere? If not, please remove.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt (hunk 5 above)
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: We should never show text to the user that's manually concatenated. Instead, we should use string interpolation with a strings.xml string so that translators can move around the sentence structure to match different locales.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragment.kt [hunk 13]
```diff
@@ -17,6 +18,7 @@ class ProfileChooserFragment : InjectableFragment() {
     fragmentComponent.inject(this)
   }
 
+  @ExperimentalCoroutinesApi
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragment.kt]: Why is this needed?

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 14]
```diff
@@ -1,29 +1,92 @@
...
+      .newBuilder<ProfileChooserModel>()
+      .registerViewTypeComputer { value ->
+        value.modelTypeCase.number
+      }
+      .registerViewDataBinderWithSameModelType(
+        viewType = ProfileChooserModel.PROFILE_FIELD_NUMBER,
+        inflateDataBinding = ProfileChooserProfileViewBinding::inflate,
+        setViewModel = this::bindProfileView)
+      .registerViewDataBinderWithSameModelType(
+        viewType = ProfileChooserModel.ADDPROFILE_FIELD_NUMBER,
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt]: Should be ADD_PROFILE_FIELD_NUMBER

Maybe rename the enum constant if this doesn't match.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 15]
```diff
@@ -1,29 +1,92 @@
...
+  }
+
+  private fun bindAddView(binding: ProfileChooserAddViewBinding, data: ProfileChooserModel) {
+    binding.root.setOnClickListener {
+      if (getAdminAuthFragment() == null) {
+        fragment.requireActivity().supportFragmentManager.beginTransaction()
+          .setCustomAnimations(R.anim.slide_up, R.anim.slide_down, R.anim.slide_up, R.anim.slide_down).add(
+          R.id.profile_chooser_fragment_placeholder,
+          AdminAuthFragment()
+        ).addToBackStack(null).commit()
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt]: Can we use an activity for this, instead, to avoid backstack management? Or, can we just set up the fragment placeholder to be on top of all other views in this activity and use a regular commitNow() + set the placeholder to visible?

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 16]
```diff
@@ -1,29 +1,92 @@
...
 /** The presenter for [ProfileChooserFragment]. */
 @FragmentScope
 class ProfileChooserFragmentPresenter @Inject constructor(
   private val fragment: Fragment,
-  private val viewModelProvider: ViewModelProvider<ProfileChooserViewModel>
+  private val viewModelProvider: ViewModelProvider<ProfileChooserViewModel>,
+  private val profileManagementController: ProfileManagementController
 ) {
+  /** Binds ViewModel and sets up RecyclerView Adapter. */
+  @ExperimentalCoroutinesApi
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt]: Why is this needed?

File: app/src/main/java/org/oppia/app/profile/ProfileChooserViewModel.kt [hunk 17]
```diff
@@ -1,10 +1,52 @@
...
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.Transformations
 import org.oppia.app.fragment.FragmentScope
+import org.oppia.app.model.Profile
+import org.oppia.app.model.ProfileChooserModel
 import org.oppia.app.viewmodel.ObservableViewModel
+import org.oppia.domain.profile.ProfileManagementController
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.logging.Logger
+import java.util.*
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserViewModel.kt]: Ditto for this import & elsewhere.

File: app/src/main/res/layout/admin_auth_fragment.xml [hunk 18]
```diff
@@ -1,7 +1,13 @@
...
   <androidx.constraintlayout.widget.ConstraintLayout
+    android:id="@+id/admin_auth_container"
     android:layout_width="match_parent"
-    android:layout_height="match_parent"></androidx.constraintlayout.widget.ConstraintLayout>
+    android:layout_height="match_parent"
+    android:background="@color/white">
+    <TextView
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      android:text="Admin Auth" />
```
Comment (BenHenning) [inline app/src/main/res/layout/admin_auth_fragment.xml:11]: Please move to strings.xml.

File: app/src/main/res/values/colors.xml (hunk 9 above)
Comment (BenHenning) [reply app/src/main/res/values/colors.xml]: +1

File: model/src/main/proto/profile.proto [hunk 19]
```diff
@@ -52,3 +52,10 @@ message ProfileId {
    */
   int32 internal_id = 1;
 }
+
+message ProfileChooserModel {
```
Comment (BenHenning) [inline model/src/main/proto/profile.proto:57]: Please document this & its fields.

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt [hunk 20]
```diff
@@ -0,0 +1,33 @@
+package org.oppia.domain.profile
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import org.oppia.app.model.ProfileId
+import javax.inject.Inject
+
+class ProfileTestHelper @Inject constructor(
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt:7]: Please add KDoc for this helper to explain when it should be used &  how.

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    profileManagementController.addProfile("Sean", "12345", null, allowDownloadAccess = true, isAdmin = true)
+    profileManagementController.addProfile("Ben", "123", null, allowDownloadAccess = false, isAdmin = false)
+    profileManagementController.loginToProfile(ProfileId.newBuilder().setInternalId(0).build())
+  }
+
+  /** Create [numProfiles] number of user profiles. */
+  @ExperimentalCoroutinesApi
+  fun addMoreProfiles(numProfiles: Int) {
+    for (x in 0 until numProfiles) {
+      profileManagementController.addProfile((x+65).toChar().toString(), "123", null, allowDownloadAccess = false, isAdmin = false)
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt]: Nit: ``(x + 65)``

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import org.oppia.app.model.ProfileId
+import javax.inject.Inject
+
+class ProfileTestHelper @Inject constructor(
+  private val profileManagementController: ProfileManagementController
+) {
+  /** Creates one admin profile and one user profile. Logs in to admin profile. */
+  @ExperimentalCoroutinesApi
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt]: Here & elsewhere in this helper: is this actually needed?

File: domain/src/test/java/org/oppia/domain/profile/ProfileTestHelperTest.kt [hunk 23]
```diff
@@ -0,0 +1,230 @@
...
+    advanceUntilIdle()
+    profileManagementController.getProfiles().observeForever(mockProfilesObserver)
+    advanceUntilIdle()
+
+    verify(mockProfilesObserver, atLeastOnce()).onChanged(profilesResultCaptor.capture())
+    assertThat(profilesResultCaptor.value.isSuccess()).isTrue()
+    assertThat(profilesResultCaptor.value.getOrThrow().size).isEqualTo(10)
+  }
+
+
```
Comment (BenHenning) [inline domain/src/test/java/org/oppia/domain/profile/ProfileTestHelperTest.kt]: Nit: extra newline

File: app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt [hunk 24]
```diff
@@ -0,0 +1,153 @@
...
+import org.oppia.util.logging.GlobalLogLevel
+import org.oppia.util.logging.LogLevel
+import org.oppia.util.threading.BackgroundDispatcher
+import org.oppia.util.threading.BlockingDispatcher
+import javax.inject.Inject
+import javax.inject.Qualifier
+import javax.inject.Singleton
+
+@RunWith(AndroidJUnit4::class)
+class ProfileChooserFragmentTest {
```
Comment (BenHenning) [inline app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:40]: Also: can we test that selecting a profile correctly routes to the home activity?

What about the tests for the auth flows?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jamesxu0!

File: app/build.gradle (hunk 1 above)
Comment (jamesxu0) [reply app/build.gradle]: Updated.

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (jamesxu0) [reply app/src/main/AndroidManifest.xml]: Updated.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt (hunk 6 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Removed.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt (hunk 12 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Updated to only import Date.

File: app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt (hunk 3 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt:76]: I think it's needed for the gravatar image loading.

File: app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt (hunk 11 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/ImageViewBindingAdapters.kt]: Removed PNG and added a vector asset.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt (hunk 4 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Removed file.

Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Removed file.

File: app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt (hunk 5 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Removed file.

Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/databinding/TextViewBindingAdapters.kt]: Removed file.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragment.kt (hunk 13 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragment.kt]: Removed.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 7 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:23]: Reorganized.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 16 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt]: Removed.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 15 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt]: Both AdminAuthFragment and AddProfileFragment are changed to activities in the next PR.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserViewModel.kt (hunk 17 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserViewModel.kt]: Updated to only import Locale.

File: app/src/main/java/org/oppia/app/splash/SplashActivity.kt (hunk 8 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/splash/SplashActivity.kt]: Updated.

File: app/src/main/res/layout/admin_auth_fragment.xml (hunk 18 above)
Comment (jamesxu0) [reply app/src/main/res/layout/admin_auth_fragment.xml:11]: This is temporary text that is removed in the next PR.

File: app/src/main/res/values/colors.xml (hunk 9 above)
Comment (jamesxu0) [reply app/src/main/res/values/colors.xml]: Updated.

File: app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt (hunk 24 above)
Comment (jamesxu0) [reply app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:40]: These tests are added in later a PR.

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt (hunk 22 above)
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt]: This is needed because addProfile uses deferred.getCompleted() which needs to be marked by @ExperimentalCoroutinesApi.

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt (hunk 21 above)
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt]: Updated.

File: domain/src/test/java/org/oppia/domain/profile/ProfileTestHelperTest.kt (hunk 23 above)
Comment (jamesxu0) [reply domain/src/test/java/org/oppia/domain/profile/ProfileTestHelperTest.kt]: Updated.

File: model/src/main/proto/profile.proto (hunk 19 above)
Comment (jamesxu0) [reply model/src/main/proto/profile.proto:57]: Added comments.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 14 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt]: Updated.

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt (hunk 20 above)
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt:7]: Added KDoc.

File: app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt (hunk 10 above)
Comment (jamesxu0) [reply app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:72]: Updated to not use coroutines anymore. Please let me know if the pass test for you now because it works for me locally.

**Comment by jamesxu0** [comment]
@BenHenning @rt4914 PTAL. I addressed all comments and fixed the test case.

File: domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt (hunk 22 above)
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/profile/ProfileTestHelper.kt]: Nevermind, this is no longer needed after I changed ProfileManagementController.

File: app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+      onView(atPositionOnView(R.id.profile_recycler_view, 0, R.id.profile_name_text)).check(matches(withText("Sean")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 1, R.id.profile_name_text)).check(matches(withText("A")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 2, R.id.profile_name_text)).check(matches(withText("B")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 3, R.id.profile_name_text)).check(matches(withText("Ben")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 4, R.id.profile_name_text)).check(matches(withText("C")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 5, R.id.profile_name_text)).check(matches(withText("D")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 6, R.id.profile_name_text)).check(matches(withText("E")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 7, R.id.profile_name_text)).check(matches(withText("F")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 8, R.id.profile_name_text)).check(matches(withText("G")))
+      onView(atPositionOnView(R.id.profile_recycler_view, 9, R.id.profile_name_text)).check(matches(withText("H")))
```
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:84]: This test case is failing, just because of one small point.
Add this `onView(withId(R.id.profile_recycler_view)).perform(scrollToPosition<RecyclerView.ViewHolder>(8))`

Meaning, you need to scroll to a particular item before writing test for it.

Screenshot 2019-11-29 at 12 15 18 AM

**Review by rt4914 — APPROVED** [review]
One test case issue which needs nit change.

File: model/src/main/proto/profile.proto [hunk 26]
```diff
@@ -52,3 +52,13 @@ message ProfileId {
...
   int32 internal_id = 1;
 }
+
+// Used in BindableAdapter for ProfileChooserFragment.
+message ProfileChooserModel {
+  oneof model_type {
+    // Indicates an individual Profile.
+    Profile profile = 1;
+    //Indicates the add profile button.
+    bool addProfile = 2;
```
Comment (BenHenning) [inline model/src/main/proto/profile.proto:62]: Should be ``add_profile``.

This will lead to changes in Kotlin, as well (``ADD_PROFILE`` will be the constant, instead).

File: model/src/main/proto/profile.proto [hunk 27]
```diff
@@ -52,3 +52,13 @@ message ProfileId {
    */
   int32 internal_id = 1;
 }
+
+// Used in BindableAdapter for ProfileChooserFragment.
+message ProfileChooserModel {
```
Comment (BenHenning) [inline model/src/main/proto/profile.proto:57]: Suggest: ProfileChooserUiModel to indicate this is a UI model.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 28]
```diff
@@ -1,29 +1,97 @@
...
+          ).add(
+            R.id.profile_chooser_fragment_placeholder,
+            AdminAuthFragment()
+          ).addToBackStack(null).commit()
+      }
+    }
+  }
+
+  private fun getAdminAuthFragment(): AdminAuthFragment? {
+    return fragment.requireActivity().supportFragmentManager.findFragmentById(R.id.profile_chooser_fragment_placeholder) as? AdminAuthFragment?
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:95]: Optional: I don't think the ``?`` is needed at the end of ``AdminAuthFragment`` here since you're using ``as?``

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 29]
```diff
@@ -1,29 +1,97 @@
...
+            .startActivity(Intent(fragment.context, HomeActivity::class.java))
+        }
+      })
+    }
+  }
+
+  private fun bindAddView(binding: ProfileChooserAddViewBinding, data: ProfileChooserModel) {
+    binding.root.setOnClickListener {
+      if (getAdminAuthFragment() == null) {
+        fragment.requireActivity().supportFragmentManager.beginTransaction()
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:80]: You could alternatively inject the activity rather than calling requireActivity().

That being said, it's a bit dubious to be changing activity-level children within a fragment under that activity. We should be signaling up to the activity that something needs to happen (e.g. via a listener), then rely on the activity to make the fragment transaction.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt [hunk 30]
```diff
@@ -1,29 +1,97 @@
...
+          .setCustomAnimations(
+            R.anim.slide_up,
+            R.anim.slide_down,
+            R.anim.slide_up,
+            R.anim.slide_down
+          ).add(
+            R.id.profile_chooser_fragment_placeholder,
+            AdminAuthFragment()
+          ).addToBackStack(null).commit()
```
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:89]: Always use commitNow(). Are we removing the back stack behavior in the next PR? Please add a TODO either way since we want to avoid using the fragment backstack in favor of using activities for cleaner transitions.

File: app/src/main/res/layout/profile_chooser_fragment.xml [hunk 31]
```diff
@@ -1,14 +1,42 @@
...
+    android:layout_height="match_parent"
+    android:background="@color/profileChooserBackground">
+    <TextView
+      android:id="@+id/profile_select_text"
+      android:layout_width="wrap_content"
+      android:layout_height="wrap_content"
+      app:layout_constraintTop_toTopOf="parent"
+      app:layout_constraintStart_toStartOf="parent"
+      android:text="@string/profile_chooser_select"
+      android:layout_marginStart="25dp"
```
Comment (BenHenning) [inline app/src/main/res/layout/profile_chooser_fragment.xml]: Here & elsewhere: we should be using dimensions that are multiples of 4dp.

**Review by BenHenning — APPROVED** [review]
Thanks @jamesxu0. This LGTM except the bit about changing the activity's children within a child. If that's going to go away in a subsequent PR, please feel free to ignore. If not, please fix here & re-assign to me if you think it's necessary. Otherwise, feel free to merge this in once comments are resolved.

File: app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt (hunk 25 above)
Comment (jamesxu0) [reply app/src/sharedTest/java/org/oppia/app/profile/ProfileChooserFragmentTest.kt:84]: Updated to include scrolls.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 29 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:80]: Since the fragment transactions are temporary for this PR, I will update to inject activity in PR #327.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 30 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:89]: This is removed in PR #327 when the fragments are switched to activites.

File: app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt (hunk 28 above)
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/profile/ProfileChooserFragmentPresenter.kt:95]: This function is removed in PR #327.

File: model/src/main/proto/profile.proto (hunk 27 above)
Comment (jamesxu0) [reply model/src/main/proto/profile.proto:57]: Updated.

File: model/src/main/proto/profile.proto (hunk 26 above)
Comment (jamesxu0) [reply model/src/main/proto/profile.proto:62]: Updated.

File: app/src/main/res/layout/profile_chooser_fragment.xml (hunk 31 above)
Comment (jamesxu0) [reply app/src/main/res/layout/profile_chooser_fragment.xml]: Updated to multiples of 4.
